### ISSUE TITLE
No possible way to set dtr and rts state during method open.

### ISSUE THREAD

**Issue by gitcnd** [pyserial/pyserial#729]
This **works**, and properly sets the pin state:-

```
import serial
ser = serial.Serial()
ser.dtr=0
ser.rts=0
ser.port='/dev/ttyS9'
ser.baudrate=115200
ser.rtscts=0
ser.dsrdtr=0
ser.open() # Works
```

However, the **below does not work**, and is the closest possible method doing the same as the above, except, it does not allow any users to set dtr/rts state (both default to 1 on windows, which generates unwanted MCU resets when trying to connect to arduino/esp32/etc devices)

`ser = serial.Serial('/dev/ttyS9', baudrate=115200, interCharTimeout=1, rtscts=0, dsrdtr=0) # fails (sets pins unwantedly)`

The above method needs to add "dtr" and "rts" kwargs to allow users to specify the initial state for those pins.

**Comment by fluffy** [comment]
In the above code, where you set the ser.dtr etc, when I look at the output of the DTR and RTS pins on a max connected to a CH340 USB to serial device, I still see the DTR and RTS go high for 20 ms when the port opens before they get forced low. So the problem might be bigger than just missing in the open function call.

**Comment by gitcnd** [comment]
Try a second time - I notice the first call when you reboot your PC (or connect a new device) does that.  From then on, it behaves.
I'm using WSL1 on Windows 10N on a RoG laptop, and CH340USB ESP32CAM - the above fix solved all my rebooting and lockup problems.

I notice the code for mpremote and pyboard on micropython already implements the same workaround when they detect windows (with a comment about ES32 and ESP8266 RTS issues), so it appears to also have solved the problem for other people on other devices as well.

On a philosophical topic instead now though: since there does exist dtr and rts state-setting "methods", it does seem to be an obvious oversight that someone forgot to allow those to be set inside the constructor.  The code which opens the port has to make a decision about whether "high" or "low" should be the initial state for those pins - but without any way for us to tell it.

**Comment by fluffy** [comment]
I get what you are saying about constructor has to choose something.  I'm looking at the posix code that is used on mac, linux, unix. I'm not how different the windows is as I don't have an easy way to test that. I could be wrong but it looks like the code has to open a socket, then it can call an ioctl on that socket to set the values of RTS / DTR. As soon as the sockets opens, the lines seem to go high. I can't see any way of setting them before the open call. The ioctl can then set them low or high. When the socket is closed, it looks like the default behavior is to set them low but it might be the case that if the "HUPCL"  ( no hangup on close ) options is set correctly that they can stay high after the socket closes. This would keep them from going low again when the device gets opened again. 

I know on the ESP32, most people use a two transistor circuit for this. You can see it in the thread on https://forum.micropython.org/viewtopic.php?t=4607 

Anyways, just poking around on this and trying to sort it out and I hope there is some way to control the RTS and DTR when the open happens but no idea what it is.

**Comment by gitcnd** [comment]
I'm in Ubuntu under WSL1 on Windows, so I wonder if my experience is different to plain linux ?

Can you explain what you mean by "socket" ?  (Did you mean "Device" instead?).  Here's what perplexity thinks:

https://www.perplexity.ai/search/C-code-in-4ELrwOKlQ1yLK42leua5Dw?s=c

**Comment by gitcnd** [comment]
I did some more serious research. 

TLDR; setting ser.rtscts=True causes the DTR line to behave (meaning - python does not pull the line high or low without us wanting it to).  
Unfortunately, nothing causes the RTS line to behave (it pulls either high or low - whatever is the opposite of what rtscts was set to) without any obvious way to tell python to leave it alone.
The [Serial.rs485_mode](https://pythonhosted.org/pyserial/pyserial_api.html?highlight=rtscts#serial.Serial.rs485_mode) looks suspiciously likely to be the culprit: that code forces the RTS state to be set (either high or low) without any option to "not touch it"... perhaps some logic from that has leaked out into the main code?  Alternatively, this doc line looks highly suspect: "_dsrdtr is not supported by all platforms (silently ignored). Setting it to None has the effect that its state follows rtscts_." ... this means that it is NOT a "bool" anymore, but a tri-state, and this quirk is not mentioned in the "Parameters" section. It also is not a correct statement (setting it to None is the same as setting it to False - it does not "follow rtscts" at all), and the logic looks bad: if something is "not supported", it should be left alone, but if it really is "not supported", how is it possible for its state to follow rtscts?  you can't set the state of something that's not supported - if you could set it... then it *is* supported...

Below is some "C" code which simply displays the current state of the serial port:

$ cat serial_info.c
```C
#include <stdio.h>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <serial_device>\n", argv[0]);
        return 1;
    }

    const char *device = argv[1];
    int fd;
    struct termios options;
    int status;

    fd = open(device, O_RDONLY | O_NOCTTY);
    if (fd == -1) {
        perror("open_port: Unable to open device");
        return 1;
    }

    tcgetattr(fd, &options);
    printf("Serial port: %s\t",argv[1]);
    printf("Baud rate: %d\t", cfgetispeed(&options));
    printf("Data bits: %d\t", (options.c_cflag & CSIZE) == CS8 ? 8 : (options.c_cflag & CSIZE) == CS7 ? 7 : 6);
    printf("Parity: %s\t", (options.c_cflag & PARENB) ? "Enabled" : "Disabled");
    printf("Stop bits: %s\t", (options.c_cflag & CSTOPB) ? "2" : "1");

    ioctl(fd, TIOCMGET, &status);
    printf("DTR: %s\t", (status & TIOCM_DTR) ? "On " : "Off");
    printf("RTS: %s\n", (status & TIOCM_RTS) ? "On " : "Off");

    close(fd);
    return 0;
}
```

Below is some python which attempts to do exactly the same thing, but fails:

$ cat serial_info.py
```python
#!/usr/bin/env python3
  
import sys
import serial
import termios
import fcntl
import struct

def get_dtr_rts_state(fd):
    buf = fcntl.ioctl(fd, termios.TIOCMGET, struct.pack('I', 0))
    status = struct.unpack('I', buf)[0]
    dtr = bool(status & termios.TIOCM_DTR)
    rts = bool(status & termios.TIOCM_RTS)
    return dtr, rts

def set_dtr_rts_state(fd, dtr, rts):
    TIOCMSET = 0x5418  # Linux-specific constant
    status = 0
    if dtr:
        status |= termios.TIOCM_DTR
    if rts:
        status |= termios.TIOCM_RTS
    fcntl.ioctl(fd, TIOCMSET, struct.pack('I', status))

def get_serial_info(port):
    ser = serial.Serial()
    #ser.rts = False
    #ser.dtr = 0
    ser.port = port
    ser.xonxoff = True
    #ser.rs485_mode = serial.rs485.RS485Settings(True,False) # not supported on my platform
    #ser.baudrate = baudrate
    ser.rtscts = False # default is false. State of the RTS pin will always be the opposite of this
    ser.dsrdtr = True # default is false. Setting this to true causes DTR to work OK
    #ser.inter_byte_timeout = 1
    #ser.exclusive = True
    ser.open()

    # rtscts    True    False   True    False
    # dsrdtr    True    False   False   True
    # DTR       works   *on*    *on*    works
    # RTS       *off*   *on*    *off*   *on*


    fd = ser.fileno()
    initial_dtr, initial_rts = get_dtr_rts_state(fd)

    print(f"Serial port: {port}",end="\t")
    print(f"Baud rate: {ser.baudrate}",end="\t")
    print(f"Data bits: {ser.bytesize}",end="\t")
    print(f"Parity: {'Enabled' if ser.parity != serial.PARITY_NONE else 'Disabled'}",end="\t")
    print(f"Stop bits: {ser.stopbits}",end="\t")

    current_dtr, current_rts = get_dtr_rts_state(fd)
    print(f"DTR: {'On ' if current_dtr else 'Off'}",end="\t")
    print(f"RTS: {'On ' if current_rts else 'Off'}",end="\t")

    # Restore initial DTR and RTS state
    set_dtr_rts_state(fd, initial_dtr, initial_rts)

    ser.close()

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <serial_device>")
        sys.exit(1)

    port_name = sys.argv[1]
    get_serial_info(port_name)
```

I assembled nice a collection of 4 MCU boards - every one of which behaves differently; an ESP32_CAM (DTR/RTS wired directly to GPIO0/Reset), a "CYB" (ESP32 wired like the ESP32_CAM, except with a cap on the Reset - so it resets, but holding it low doesn't keep it stuck in reset mode like the CAM), an ESP32_C3 (does not use hardware flow control, can be hardware-reset by manipulating DTR/RTS lines) and and ESP32_S2 (uses hardware flow control, cannot be hardware-reset)

ttyS10=CAM
ttyS26=CYB 
ttyS23=LOLIN_C3_MINI 
ttyS12=ESP32S3 

I ran this test, which 
1. sets the hadrware state of the serial port lines (perl)
2. reads the state in "C" code, which always correctl reflects (1) above
3. reads the state in "python" using pyserial - which always messes up the RTS line, and 50% of the time messes up DTR as well

    for PORT in /dev/ttyS12 /dev/ttyS23 /dev/ttyS10 /dev/ttyS26;do echo -e "\n" $GRN $PORT $NORM;perl ~/Downloads/mcu_serial.pl -port $PORT -setdtr -setrts -exit;./serial_info $PORT ; echo -n $YEL; ./serial_info.py $PORT ;echo $NORM; perl ~/Downloads/mcu_serial.pl -port $PORT -clrdtr -clrrts -exit;./serial_info $PORT ; echo -n $YEL; ./serial_info.py $PORT ;echo $NORM; done

Output looks like:-

```
  /dev/ttyS12 

Serial port: /dev/ttyS12        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off
Serial port: /dev/ttyS12        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS12        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
Serial port: /dev/ttyS12        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off

  /dev/ttyS23 

Serial port: /dev/ttyS23        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: On         RTS: Off
Serial port: /dev/ttyS23        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS23        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: Off        RTS: Off
Serial port: /dev/ttyS23        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off

  /dev/ttyS10 

Serial port: /dev/ttyS10        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off
Serial port: /dev/ttyS10        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS10        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
Serial port: /dev/ttyS10        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off

  /dev/ttyS26 

Serial port: /dev/ttyS26        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off
Serial port: /dev/ttyS26        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS26        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
Serial port: /dev/ttyS26        Baud rate: 9600 Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
```

(no matter how I adjust the pyserial settings, there is always 1MCU in the above 4 that remains unusable, and usually 1, 2, or 3 which suffer unwanted hardware resets as a side-effect of simply opening the serial port)

so through exhaustive trial-and-error, I can confidently say that pyserial has a bug in the way it handles the RST line **for sure.**

There is no work-around for this bug.  No possible combination of initial settings make all microcontroller boards work (we either cannot talk to anything with hardware flow control, or, we cause unwanted resets on the other boards).

The fix here is to change pyserial so it does **not** alter the sate of the DTR/RTS pins when the user has not specific hardware flow control.  The user turned that off for a reason, so it makes sense to respect their decision and leave those pins alone, since you know for-sure (since they turned it off) that they're not used for flow control, and you cannot know what they are used for.

```
$ uname -a
Linux RoG 4.4.0-17763-Microsoft #2268-Microsoft Thu Oct 07 16:36:00 PST 2021 x86_64 x86_64 x86_64 GNU/Linux

```

**Comment by gitcnd** [comment]
p.s. "not specific hardware flow control" means:-

    xonxoff = True
    rtscts = False

**Comment by gitcnd** [comment]
Here is a python program which **does** work properly. It does not use pyserial:

$ cat  serial_info5.py 
```python
#!/usr/bin/env python3

import sys
import os
import fcntl
import termios
import struct

def get_dtr_rts_state(fd):
    TIOCMGET = termios.TIOCMGET
    buf = fcntl.ioctl(fd, TIOCMGET, struct.pack('I', 0))
    status = struct.unpack('I', buf)[0]
    dtr = bool(status & termios.TIOCM_DTR)
    rts = bool(status & termios.TIOCM_RTS)
    return dtr, rts

def get_serial_info(port):
    try:
        fd = os.open(port, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)
    except OSError as e:
        print(f"Error opening serial port {port}: {e}")
        return

    # Using termios.TCGETS to get the current terminal attributes
    TCGETS = termios.TCGETS
    buf = struct.pack('I'*32, *([0]*32))
    attrs = fcntl.ioctl(fd, TCGETS, buf)
    attrs = struct.unpack('I'*32, attrs)

    # Baud rate
    baud_rate = attrs[4]  # Input baud rate

    # Data bits
    data_bits = attrs[2] & termios.CSIZE
    if data_bits == termios.CS8:
        data_bits = 8
    elif data_bits == termios.CS7:
        data_bits = 7
    elif data_bits == termios.CS6:
        data_bits = 6
    elif data_bits == termios.CS5:
        data_bits = 5
    else:
        data_bits = None

    # Parity
    parity = "Enabled" if attrs[2] & termios.PARENB else "Disabled"

    # Stop bits
    stop_bits = 2 if attrs[2] & termios.CSTOPB else 1

    # Get DTR and RTS state
    dtr, rts = get_dtr_rts_state(fd)

    baud_rate=11

    # Print the serial port info
    print(f"Serial port: {port}", end="\t")
    print(f"Baud rate: {baud_rate}", end="\t")
    print(f"Data bits: {data_bits}", end="\t")
    print(f"Parity: {parity}", end="\t")
    print(f"Stop bits: {stop_bits}", end="\t")
    print(f"DTR: {'On ' if dtr else 'Off'}", end="\t")
    print(f"RTS: {'On ' if rts else 'Off'}", end="\t")

    # Close the file descriptor
    os.close(fd)

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <serial_device>")
        sys.exit(1)

    port_name = sys.argv[1]
    get_serial_info(port_name)
```

You can see, in every test case, that the state of the DTR/RTS pins prior to the python program running, are accurately preserved and reflected from the python program:

```
for PORT in /dev/ttyS12 /dev/ttyS23 /dev/ttyS10 /dev/ttyS26; do echo -e "\n" $GRN $PORT $NORM;for RTS in setrts clrrts;do for DTR in setdtr clrdtr;do perl ~/Downloads/mcu_serial.pl -port $PORT -$DTR -$RTS -exit;./serial_info $PORT ; echo -n $YEL; ./serial_info5.py $PORT ;echo $NORM; done; done; done

  /dev/ttyS12 

Serial port: /dev/ttyS12        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: On 
Serial port: /dev/ttyS12        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: On 

Serial port: /dev/ttyS12        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: On 
Serial port: /dev/ttyS12        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: On 

Serial port: /dev/ttyS12        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off
Serial port: /dev/ttyS12        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS12        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
Serial port: /dev/ttyS12        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off

  /dev/ttyS23 

Serial port: /dev/ttyS23        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: On         RTS: On 
Serial port: /dev/ttyS23        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: On         RTS: On 

Serial port: /dev/ttyS23        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: Off        RTS: On 
Serial port: /dev/ttyS23        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: Off        RTS: On 

Serial port: /dev/ttyS23        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: On         RTS: Off
Serial port: /dev/ttyS23        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: On         RTS: Off

Serial port: /dev/ttyS23        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: Off        RTS: Off
Serial port: /dev/ttyS23        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 2    DTR: Off        RTS: Off

  /dev/ttyS10 

Serial port: /dev/ttyS10        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: On 
Serial port: /dev/ttyS10        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: On 

Serial port: /dev/ttyS10        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: On 
Serial port: /dev/ttyS10        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: On 

Serial port: /dev/ttyS10        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off
Serial port: /dev/ttyS10        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS10        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
Serial port: /dev/ttyS10        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off

  /dev/ttyS26 

Serial port: /dev/ttyS26        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: On 
Serial port: /dev/ttyS26        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: On 

Serial port: /dev/ttyS26        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: On 
Serial port: /dev/ttyS26        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: On 

Serial port: /dev/ttyS26        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off
Serial port: /dev/ttyS26        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: On         RTS: Off

Serial port: /dev/ttyS26        Baud rate: 13   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
Serial port: /dev/ttyS26        Baud rate: 11   Data bits: 8    Parity: Disabled        Stop bits: 1    DTR: Off        RTS: Off
```

**Comment by fluffy** [comment]
I greatly appreciate you digging into this and tracking it all down.

**Comment by gitcnd** [comment]
I spent many more hours in Windows 10 trying to make a working version for that platform as well... progress is exhausting and so far without luck.  I'm working with C, Perl, Python, and even the cmd.exe "mode" program - so multiple different underlying implementations ... there are glitches and gotchas in all of them so far.

I might have to dust off my brain and remember how to program serial directly from assembler, which I did a few times several decades ago...

I have not even started looking at any Mac O/S (and I'd bet that the new silicon behaves differently...) nor have I investigated native linux (I'm using WSL1 here - which offers native serial access) and I'm not even sure if WSL2 can use serial ports (I know is does not have *native* access)...

So yeah - not even half way yet.
